File: apo-pot/src/main/java/com/uqbar/aop/transaction/NullObjectTransaction.java
package com.uqbar.aop.transaction;

import java.util.Map;

import com.uqbar.common.transaction.ObjectTransaction;
import com.uqbar.common.transaction.TaskOwner;

/**
 * Represents a non-transactional context.
 * 
 * @author npasserini
 * @author jfernandes
 */
public class NullObjectTransaction extends ObjectTransactionImpl {

	public NullObjectTransaction(ObjectTransactionImpl parent, TaskOwner owner, Long id) {
		super(parent, owner, id);
	}

	/**
	 * Interceptor for field reading.
	 * @return the parent transaction's value for that field, if it has a parent. Otherwise the real object's field value.
	 */
	public Object fieldRead(Object owner, String fieldName, Object value) {
		return this.getParent() != null ? this.fieldReadFromParent(owner, fieldName, value) : value;
	}

	protected Object fieldReadFromParent(Object owner, String fieldName, Object value) {
		return ((ObjectTransactionImpl) this.getParent()).fieldRead(owner, fieldName, value);
	}

	/**
	 * Interceptor for field writing. It should not be happening. You should not be modifying objects without
	 * a transaction.
	 */
	public Object fieldWrite(Object owner, String fieldName, Object newValue, Object oldValue) {
		return newValue;
	}

	public Object arrayRead(Object owner, String fieldName, Object value) {
		return value;
	}

	/**
	 * @return null, nadie puede ser padre de una NullTransaction
	 */
	protected ObjectTransaction commit() {
		this.getLogger().debug("Commit on NullTransaccion [" + this.getId() + "]");
		ObjectTransaction transaction = this.removeParent();
		this.getLogger().debug("Next transaction: [" + (transaction != null ? transaction.getId() : null) + "]");
		this.setState(ObjectTransactionImpl.STATE_COMMITED);
		return transaction;
	}

	/**
	 * Populates all values from the map to the real objects.
	 */
	protected void commitMap(Map<IdentityWrapper, Map<String, Object>> attributeMap) {
		super.populateValuesToObjects();
	}

	public void cloneAttributes(Object sourceObject, Object targetObject) {
		// nothing to be done.
	}
	
	@Override
	protected void addParentChild() {
	}
}

File: apo-pot/src/main/scala/com/uqbar/aop/transaction/NullObjectTransaction.scala
package com.uqbar.aop.transaction;

import java.util.List;
import java.util.Map;
import java.util.Set;

import com.uqbar.common.transaction.ObjectTransaction;
import com.uqbar.common.transaction.TaskOwner;
import com.uqbar.common.transaction.Collection.TransactionalData;

/**
 * Represents a non-transactional context.
 * 
 * @author npasserini
 * @author jfernandes
 */
class NullObjectTransaction(parent:ObjectTransactionImpl, owner:TaskOwner, id:Long) extends ObjectTransactionImpl(parent, owner, id){

	/**
	 * Interceptor for field reading.
	 * @return the parent transaction's value for that field, if it has a parent. Otherwise the real object's field value.
	 */
	override def fieldRead(owner:Any, fieldName:String , value:Any):Any= {
		if(this.parent != null) this.fieldReadFromParent(owner, fieldName, value) else value
	}

	protected def fieldReadFromParent(owner:Any, fieldName:String , value:Any):Any= {
		return parent.fieldRead(owner, fieldName, value)
	}

	/**
	 * Interceptor for field writing. It should not be happening. You should not be modifying objects without
	 * a transaction.
	 */
	override def doFieldWrite[T](owner:Object, fieldName:String, newValue:T, oldValue:T): T={
		return newValue;
	}

	/**
	 * @return null, nadie puede ser padre de una NullTransaction
	 */
	override protected def commit():ObjectTransaction = {
		logger.debug("Commit on NullTransaccion [" + this.getId() + "]");
		val transaction = this.removeParent();
		logger.debug("Next transaction: [" + (if(transaction != null) transaction.getId() else null) + "]");
		state = ObjectTransactionImpl.STATE_COMMITED
		transaction;
	}

	/**
	 * Populates all values from the map to the real objects.
	 */
	protected def commitMap(attributeMap:Map[IdentityWrapper, Map[String, Any]]) {
		super.populateValuesToObjects();
	}

	override def cloneAttributes(sourceObject:Object, targetObject:Object) {	}
	
	override protected def addParentChild(){}
}